server/polar/subscription/service.py

```diff
@@ -783,26 +800,90 @@ async def update_product(
-                product.organization_id
-            )
-            assert organization is not None
             proration_behavior = organization.proration_behavior
 
-        await stripe_service.update_subscription_price(
-            subscription.stripe_subscription_id,
-            new_prices=[
-                price.stripe_price_id for price in prices if is_static_price(price)
-            ],
-            proration_behavior=proration_behavior.to_stripe(),
-            metadata={
-                "type": ProductType.product,
-                "product_id": str(product.id),
-            },
-        )
+        if organization.subscriptions_billing_engine is False:
+            # Stripe behavior
+            if subscription.stripe_subscription_id is None:
+                raise SubscriptionNotActiveOnStripe(subscription)
+
+            await stripe_service.update_subscription_price(
+                subscription.stripe_subscription_id,
+                new_prices=[
+                    price.stripe_price_id for price in prices if is_static_price(price)
+                ],
+                proration_behavior=proration_behavior.to_stripe(),
+                metadata={
+                    "type": ProductType.product,
+                    "product_id": str(product.id),
+                },
+            )
+        else:
+            now = datetime.now(UTC)
+            # Cycle start is assumed to be unchanged across plan/product changes (e.g. monthly to yearly)
+            cycle_start = subscription.current_period_start
+
+            # Cycle end can change in the case of e.g. monthly to yearly
+            old_cycle_end = previous_product.recurring_interval.get_next_period(
+                subscription.current_period_start
+            )
+            new_cycle_end = subscription.recurring_interval.get_next_period(
+                subscription.current_period_start
+            )
+
+            old_cycle_remaining_time = (old_cycle_end - now).total_seconds()
+            old_cycle_total_time = (old_cycle_end - cycle_start).total_seconds()
+            old_cycle_pct_remaining = old_cycle_remaining_time / old_cycle_total_time
+
+            new_cycle_remaining_time = (new_cycle_end - now).total_seconds()
+            new_cycle_total_time = (new_cycle_end - cycle_start).total_seconds()
+            new_cycle_pct_remaining = new_cycle_remaining_time / new_cycle_total_time
+
+            subscription.current_period_end = new_cycle_end
+
+            entry_unused_time = BillingEntry(
+                direction=BillingEntryDirection.credit,
+                start_timestamp=cycle_start,
+                end_timestamp=now,
+                amount=round(
+                    previous_product.prices[0].price_amount * old_cycle_pct_remaining
+                ),
+                currency=subscription.currency,
+                customer_id=subscription.customer_id,
+                product_price_id=previous_product.prices[0].id,  # FIXME
```

Comment:
A bit tedious but I think the "right" way to fix that is to loop through all the prices and add a billing entry for each "static price" (see `product/guard.py`). The risk with `prices[0]` is that a metered price could be first in the list vs. a static price.